### ISSUE TITLE
Missing required parameter USERNAME Error on Custom Auth.

### ISSUE THREAD

**Issue by mehtabghani10p** [aws-amplify/amplify-ios#1360]
### Describe the bug

Getting "Missing required parameter USERNAME" error while calling:
Amplify.Auth.signIn(username: userName)

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Sign in with Custom Auth (signIn followed by challengeResponse in confirmSignIn)
2. Then log out the user using Amplify.Auth.signOut
3. Sign in with "different user" following step 1.
4. See error

Note: Sometimes an error occurs on the first try, sometimes it takes 2 or more attempts.
```

### Expected behavior

should proceed with 'confirmSignInWithCustomChallenge' 

### Amplify Framework Version

1.13.1

### Amplify Categories

Auth

### Dependency manager

Cocoapods

### Swift version

4.2

### CLI version

not using

### Xcode version

12.5

### Device

iPhone7+

### iOS Version

14.5

### Specific to simulators

No

### Additional context

Image 1:
<img width="696" alt="MicrosoftTeams-image (1)" src="https://user-images.githubusercontent.com/84833984/128518797-c995fd85-ea1b-4ccf-9fa0-b6ee6b2df1bb.png">

Image 2:
<img width="738" alt="Screenshot 2021-08-06 at 4 00 12 PM" src="https://user-images.githubusercontent.com/84833984/128518728-6da2224e-a5af-4532-a273-118d0d09bb98.png">

As seen in screenshots userName is there (Image1 & Image2). however, API still throws an error (Image 2)

**Comment by royjit** [comment]
I was not able to reproduce the issue by following the steps you mentioned. Can you enable verbose logging for the inner SDK and send us (after redacting sensitive info) the request / response / error that you see? You can enable logs by:

```swift
AWSDDLog.sharedInstance.add(AWSDDTTYLogger())
AWSDDLog.sharedInstance.logLevel = .verbose
```

**Comment by mehtabghani10p** [comment]
@royjit logs file is attached, hope it would be helpful.
[verbose-log.txt](https://github.com/aws-amplify/amplify-ios/files/7029904/verbose-log.txt)

**Comment by mehtabghani10p** [comment]
@royjit I've attached a POC, where I'm able to reproduce this issue. 

Note: 
Before signing in, URLs are being downloaded using Amplify.Storage.getURL(key: key). (Line# 43). This depicts the scenario actually happening in a real app.
Also tried to cancel "getURL" operations before signing out but it didn't work.

Pre-requisite:
1- pod install
2- replace '*' with necessary ids/info in amplifyconfiguration.json
3- provide two different users' data in ViewController.swift (Line# 14)

Steps To Reproduce:
1- Press "Signin User1" button and wait.
2- Upon logged in successfully, Press "Signin User2" button.

Actual Behaviour:
throwing error "Missing required parameter USERNAME"

if you need further details feel free to ask.

[AmplifyPOC.zip](https://github.com/aws-amplify/amplify-ios/files/7047286/AmplifyPOC.zip)

**Comment by royjit** [comment]
I tried the same with your POC and unfortunately could not reproduce the issue. The lambda trigger I used is given below, but I do not suspect that the error is from lambda because from your logs it is clear that the app is not sending the USERNAME field.

Amplify signIn api should have thrown an error if the username is empty [CodeRef](https://github.com/aws-amplify/amplify-ios/blob/1c19d2a1770aa99b4670c6a1c7eeed386dab93c1/AmplifyPlugins/Auth/AWSCognitoAuthPlugin/Request/AuthSignInRequest%2BValidate.swift#L13). But for some reason the username was changed to an empty string.

Another place that this might have happened is [here](https://github.com/aws-amplify/amplify-ios/blob/1c19d2a1770aa99b4670c6a1c7eeed386dab93c1/AmplifyPlugins/Auth/AWSCognitoAuthPlugin/Dependency/AuthenticationProviderAdapter%2BSignIn.swift#L24), but I need your help to debug this. Can you put a break point at line 27 on the file "AuthenticationProviderAdapter+SignIn.swift" and check if the username is empty or not ?

```js
function createAuthChallenge(event) {
  
  if (event.request.challengeName === 'CUSTOM_CHALLENGE') {
      const challengeCode = "1212";
      event.response.publicChallengeParameters = {};
      event.response.privateChallengeParameters = {};
      event.response.privateChallengeParameters.answer = challengeCode;
  }
}
```

```js
function verifyAuthChallengeResponse(event) {
  if (event.request.privateChallengeParameters.answer === event.request.challengeAnswer) {
      event.response.answerCorrect = true;
  } else {
      event.response.answerCorrect = false;
  }
}
```

**Comment by mehtabghani10p** [comment]
@royjit username was present during my testing, check screenshot.

<img width="1407" alt="Screenshot 2021-09-10 at 5 13 20 PM" src="https://user-images.githubusercontent.com/84833984/132855243-e900a13c-172b-4c4a-b9ef-458756111eca.png">

**Comment by jscottanderson** [comment]
I'm also occasionally experiencing the same issue, even though I'm definitely sending through a valid username:

```
AuthError: Missing required parameter ANSWER Recovery suggestion: Make sure that the parameters passed are valid Caused by: invalidParameter
```

**Comment by royjit** [comment]
I have added this PR https://github.com/aws-amplify/amplify-ios/pull/1549 that should fix the issue, it is under review. In the mean time will you be able to try out this branch and see if that works for you? - https://github.com/aws-amplify/amplify-ios/tree/royjit.authconcurrency

**Comment by mehtabghani10p** [comment]
@royjit Hey, let me check and will get back to you.

**Comment by mehtabghani10p** [comment]
@royjit 
It seems that branch https://github.com/aws-amplify/amplify-ios/tree/royjit.authconcurrency was deleted on Nov 30, hence we were unable to test out that branch. 

We noticed that release 1.17.0 has been released and it contains a fix for same issue. We have tried this version of amplify in our POC project and it seems that the issue is resolved. However, now another issue is observed where previous user who was signed out is successfully logged in.

Steps within POC:
1. Tap signin user 1
2. Wait for success
3. Tap signin user 2
4. Tap signin user 2, before step 3 success
5. View logs (user 1 will signin and then user 2 will signin)

**Comment by brennanMKE** [comment]
@mehtabghani10p This PR has been merged to `main` and was included in the release for 1.17.0. Please try the latest version. Could you create a new issue for the new issue you are seeing? We can track that separately.

* [CHANGELOG](https://github.com/aws-amplify/amplify-ios/blob/main/CHANGELOG.md#1170-2021-12-07)
